influxdb3_write/src/wal.rs

```diff
@@ -1,44 +1,642 @@
+use crc32fast::Hasher;
+use datafusion::parquet::file::reader::Length;
+use observability_deps::tracing::{info, warn};
+use snap::read::FrameDecoder;
+use std::fmt::Debug;
+use std::{
+    fs::{File, OpenOptions},
+    io::{self, BufReader, Cursor, Read, Write},
+    mem,
+    path::PathBuf,
+};
+use thiserror::Error;
+
+/// The first bytes written into a segment file to identify it and its version.
+type FileTypeIdentifier = [u8; 8];
+const FILE_TYPE_IDENTIFIER: &[u8] = b"idb3.001";
+
+/// File extension for segment files
+const SEGMENT_FILE_EXTENSION: &str = "wal";
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("io error: {source}")]
+    Io {
+        #[from]
+        source: io::Error,
+    },
+
+    #[error("error converting wal batch to json: {source}")]
+    Json {
+        #[from]
+        source: serde_json::Error,
+    },
+
+    #[error("converting u64 to u32: {source}")]
+    TryFromU64 {
+        #[from]
+        source: std::num::TryFromIntError,
+    },
+    #[error("invalid segment file {segment_id:?} at {path:?}: {reason}")]
+    InvalidSegmentFile {
+        segment_id: SegmentId,
+        path: PathBuf,
+        reason: String,
+    },
+
+    #[error("unable to read crc for segment {segment_id:?}")]
+    UnableToReadCrc { segment_id: SegmentId },
+
+    #[error("unable to read length for segment data {segment_id:?}")]
+    UnableToReadLength { segment_id: SegmentId },
+
+    #[error("length mismatch for segment {segment_id:?}: expected {expected}, got {actual}")]
+    LengthMismatch {
+        segment_id: SegmentId,
+        expected: u32,
+        actual: u32,
+    },
+
+    #[error("checksum mismatch for segment {segment_id:?}: expected {expected}, got {actual}")]
+    ChecksumMismatch {
+        segment_id: SegmentId,
+        expected: u32,
+        actual: u32,
+    },
+
+    #[error("invalid segment file name {0:?}")]
+    InvalidSegmentFileName(String),
+
+    #[error("unable to parse segment id from file: {source}")]
+    UnableToParseSegmentId {
+        file_name: String,
+        source: std::num::ParseIntError,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 #[derive(Debug)]
 pub struct WalImpl {
-    #[allow(dead_code)]
-    path: PathBuf,
+    root: PathBuf,
 }
 
 impl WalImpl {
-    pub fn new(path: impl Into<PathBuf>) -> Self {
-        Self { path: path.into() }
+    pub fn new(path: impl Into<PathBuf>) -> Result<Self> {
+        let root = path.into();
+        info!(wal_dir=?root, "Ensuring WAL directory exists");
+        std::fs::create_dir_all(&root)?;
+
+        // ensure the directory creation is actually fsync'd so that when we create files there
+        // we don't lose them (see: https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-pillai.pdf)
+        File::open(&root)
+            .expect("should be able to open just-created directory")
+            .sync_all()
+            .expect("fsync failure");
```

Comment:
In a few places we crash on fsync failure. Is that wanted behavior? Is it because we can't expect the files on disk to be in a good state?

Reply:
In this case we can't start up without this working so it's best to just crash out.